### PR TITLE
nrf52840 802.15.4 LQI

### PR DESCRIPTION
**Pull request by tyler-potyondy**
### Pull Request Overview

This pull request updates the 15.4 network stack to pass the LQI value up the stack. This updates the `receive()` method to accept LQI as an argument. In both the nrf52840 and rf233 radios, the LQI value is found in the receive buffer following the last byte of data received. This implementation implements the LQI for the nrf52840 and not for the rf233 radio. For now, the rf233 LQI value is set to zero. 

This PR also updates the arguments passed to the userprocess. Previously, the pan, src address, and dest address were passed to the userprocess in the three available registers. The pan field has now been replaced by the lqi value as the pan field can be obtained in userland by re parsing the header. 

### Testing Strategy

This pull request was tested on the nrf52840. 

### TODO or Help Wanted

N/A

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: capsules/extra/src/ieee802154/driver.rs [hunk 1]
```diff
@@ -1070,6 +1070,7 @@ impl<'a, M: device::MacDevice<'a>> device::TxClient for RadioDriver<'a, M> {
 
 /// Encode two PAN IDs into a single usize.
 #[inline]
+#[allow(dead_code)]
```
Comment (brghena) [inline capsules/extra/src/ieee802154/driver.rs:1073]: This doesn't need to be here, right?

File: capsules/extra/src/ieee802154/device.rs [hunk 2]
```diff
@@ -134,11 +134,19 @@ pub trait RxClient {
     /// - `header`: A fully-parsed representation of the MAC header, with the
     /// caveat that the auxiliary security header is still included if the frame
     /// was previously secured.
+    /// - `lqi`: The link quality indicator of the received frame.
```
Comment (brghena) [inline capsules/extra/src/ieee802154/device.rs:137]: Is LQI actually the correct term here, or should it be RSSI?

https://microchip.my.site.com/s/article/RSSI-vs-LQI-What-is-the-difference

File: chips/nrf52840/src/ieee802154_radio.rs [hunk 3]
```diff
@@ -801,7 +801,14 @@ impl<'a> Radio<'a> {
                     // or some instance in  driver taking buffer without properly replacing).
                     let rbuf = self.rx_buf.take().unwrap();
 
-                    let frame_len = rbuf[MIMIC_PSDU_OFFSET as usize] as usize - radio::MFR_SIZE;
+                    // data buffer format: | PSDU OFFSET | FRAME | LQI |
+                    // length of received data transferred to buffer (including PSDU)
+                    let data_len = rbuf[MIMIC_PSDU_OFFSET as usize] as usize;
+
+                    // lqi is found just after the data received
+                    let lqi = rbuf[data_len];
```
Comment (brghena) [inline chips/nrf52840/src/ieee802154_radio.rs:809]: To check that I understand: for the nRF52840 this value was already there in every packet we read, we were just never parsing it out before?

**Review by brghena — COMMENTED** [review]
Generally seems fine to me. Some comments

File: capsules/extra/src/ieee802154/driver.rs (hunk 1 above)
Comment (tyler-potyondy) [reply capsules/extra/src/ieee802154/driver.rs:1073]: I modified the arguments passed with the scheduled upcall to no longer include the pan so this function is no longer called. I figured we'd want to keep this though since it may still be useful in the future?

File: capsules/extra/src/ieee802154/device.rs (hunk 2 above)
Comment (tyler-potyondy) [reply capsules/extra/src/ieee802154/device.rs:137]: Interestingly, both rf233 and nrf52840 are calling this value "LQI" in their spec even though RSSI seems more accurate. 

In the nordic case, the "LQI" is really just the median RSSI of the 3 RSSI values taken during the reception.

File: chips/nrf52840/src/ieee802154_radio.rs (hunk 3 above)
Comment (tyler-potyondy) [reply chips/nrf52840/src/ieee802154_radio.rs:809]: That is mostly correct. The radio attempts to place the lqi into the buffer. However, we did not provide a buffer large enough for this value to be placed so it was not parsed and not in the `rbuf` struct.

File: capsules/extra/src/ieee802154/device.rs (hunk 2 above)
Comment (brghena) [reply capsules/extra/src/ieee802154/device.rs:137]: I guess I'd fall on the side of going with what hardware says then, and stick with LQI.

File: capsules/extra/src/ieee802154/driver.rs (hunk 1 above)
Comment (brghena) [reply capsules/extra/src/ieee802154/driver.rs:1073]: I see. I thought that was for debugging or something.

I could go either way on it. Probably you're right to keep it.

**Review by brghena — DISMISSED** [review]

File: capsules/extra/src/ieee802154/driver.rs [hunk 4]
```diff
@@ -1180,12 +1189,11 @@ impl<'a, M: device::MacDevice<'a>> device::RxClient for RadioDriver<'a, M> {
...
                 .unwrap_or(false);
             if read_present {
-                // Encode useful parts of the header in 3 usizes
-                let pans = encode_pans(&header.dst_pan, &header.src_pan);
+                // Encode lqi and useful parts of the header in 3 usizes
                 let dst_addr = encode_address(&header.dst_addr);
                 let src_addr = encode_address(&header.src_addr);
                 kernel_data
-                    .schedule_upcall(upcall::FRAME_RECEIVED, (pans, dst_addr, src_addr))
+                    .schedule_upcall(upcall::FRAME_RECEIVED, (lqi as usize, dst_addr, src_addr))
```
Comment (bradjc) [inline capsules/extra/src/ieee802154/driver.rs:1196]: I know we didn't promise any stability here, but I'm still concerned about churn. Do you know why pan IDs and src/dst addresses where included in this upcall in the first place? They are in they packet, no? And it's not even possible to fit a 64 bit address in a 32 bit usize.

LQI is not in the packet itself, and that to me makes sense to include in the upcall arguments. I think I would rather see a new upcall be added (eg `FRAME_RECEIVED_LQI` or something) that only provides LQI. This is assuming there is not a use case for including the PAN IDs and src/dst addresses.

I think it might make sense to include other metadata in the future (for example, CRC check status for use cases where even failed crc packets are provided to userspace, or a flag indicating encrypted or not). If we find out that there is no use case for pans/dst_addr/src_addr we can just entirely drop that upcall in the future.

File: capsules/extra/src/ieee802154/driver.rs (hunk 1 above)
Comment (bradjc) [reply capsules/extra/src/ieee802154/driver.rs:1073]: Let's decide if we want this or not. If we do, then let's have two upcalls, one that is the same as today, and one that only provides lqi.

If not, let's delete these functions. Future readers aren't going to understand what these functions are for.

Comment (tyler-potyondy) [reply capsules/extra/src/ieee802154/driver.rs:1073]: We discussed this Monday in the network-wg and decided to remove these functions. Will update PR now.

File: capsules/extra/src/ieee802154/framer.rs [hunk 5]
```diff
@@ -468,13 +470,18 @@ impl<'a, M: Mac<'a>, A: AES128CCM<'a>> Framer<'a, M, A> {
...
+        frame_len: usize,
+        lqi: u8,
+    ) -> RxState {
         // Try to decode the MAC header. Three possible results can occur:
         // 1) The frame should be dropped and the buffer returned to the radio
         // 2) The frame is unsecured. We immediately expose the frame to the
         //    user and queue the buffer for returning to the radio.
         // 3) The frame needs to be unsecured.
-        let result = Header::decode(&buf[radio::PSDU_OFFSET..], false)
+        let result = Header::decode(&buf[radio::PSDU_OFFSET..(radio::PSDU_OFFSET+radio::MAX_FRAME_SIZE)], false)
```
Comment (bradjc) [inline capsules/extra/src/ieee802154/framer.rs:484]: Oh I meant to ask about these changes as well. These seem unrelated to LQI. Are they necessary?

Comment (tyler-potyondy) [reply capsules/extra/src/ieee802154/framer.rs:484]: This gets at a major pain point in the 15.4 network stack...

The short answer is that the buffer given to the radio was increased by one byte (on nrf52840 and rf233 the lqi is placed by the radio after the packet). 

Ideally, we would then only pass a buffer of the size of the radio packet (as done previously). However, we must pass the whole buffer since the buffer is given back to the radio after being passed up through the network stack (this is a major drawback of our current design). What this in reality means is that to keep a 130 byte buffer (size needed for psdu + packet + lqi), we have to pass a buffer of this size up the entire stack. Passing a slice results in the buffer "shrinking" each time a packet is received as the radio's buffer is then replaced with the now truncated slice. This is not a unique problem and is also caused friction with the PSDU offset PR a few weeks ago.

So the reason we have to now set the upper bound on the range is so that we don't pass the lqi as part of the packet to layers higher than the framer (since the buffer at the framer includes the lqi too). I'll go ahead and add a brief comment mentioning this.

**Review by bradjc — APPROVED** [review]

**Review by ppannuto — APPROVED** [review]
